src/dune_engine/action_builder.ml

```diff
@@ -504,9 +492,121 @@ struct
     and build_static_rule_deps_and_exec :
         type a. a t -> (a * Dep.Set.t) Memo.Build.t =
      fun t ->
-      let* () = Build_deps.build_deps (static_deps t).rule_deps in
-      exec t
+      let { Static_deps.rule_deps; action_deps } = static_deps t in
+      let* () = Build_deps.build_deps rule_deps in
+      let+ x, deps = exec t in
+      (x, Dep.Set.union action_deps deps)
```

Comment:
Looks like this fixes a real bug.

Reply:
Maybe that wasn't a bug because the call site (on line 489) adds the missing dependencies?

Reply:
Well, this function is exported... and then the only call site in `build_system.ml` also adds `action_deps` back :) So, I guess we should simplify that call site.

Reply:
(Line 1221.)

Reply:
AFAICT, this function was unused in the code base. When I started using it, I encountered a bug and this was the fix. Let me double check the other line.

Reply:
Alright, so clearly the code can be simplified: from `Action_builder.Make_exec`, we only ever use `build_static_rule_deps_and_exec`, which means that we only care about building rule deps + returning all dependencies. So instead of having two functions, we could have a single function that does everything in one pass.

I will prepare a separate PR for this.

Reply:
And we indeed don't need the union in build_system.ml, I removed it